### PR TITLE
Flattened descriptor sets on Metal

### LINKED ISSUE #2295: Starting a render pass is too slow on Metal
Takes about 6% of our library's time. Presumably, the main cost comes from the need to reset all the states when we begin a pass.

What we can do is use the `set_vertex_buffers` instead of `set_vertex_buffer` and such to reset all the states at once instead of filtering them and producing separate commands for each single binding. Those "batch" commands need to be added to metal-rs first, and we need to figure out a way to use them.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #2295
Based on #2298

This gives us another 2.5% performance on the benchmark. ~~However, it is really going to shine when we start binding the descriptors in bulk (hence WIP here).~~ Edit: actually, not. I'll get that shining PR separately.

PR checklist:
- [ ] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: metal
- [ ] `rustfmt` run on changed code

### CODE REVIEW

**Comment by kvark** [comment]
Here is what this is about. We have descriptors following this route:
  1. `Device::create_descriptor_set_layout`
  2. `Device::create_pipeline_layout`
  3. `DescriptorPool::allocate_set`
  4. `Device::write_descriptor_sets`
  5. `CommandBuffer::bind_graphics_descriptor_sets`

Originally, we were keeping a hold of the set layout up until the last stage. This caused issues with spilling on writes (4) and wasn't super efficient on the binding side (5). Then, semi-recently, we started processing the descriptor layouts and sorting them in the binding order. This made the writes (4) trivial and improved the binding a bit. But essentially, the split of descriptor data into shader stages was still happening at binding (5).

What this PR does is moving the stage split earlier - to the allocation and writes (2-4). This makes the size of descriptor pools a little bit bigger, since they can hold the same descriptor multiple times (one per shader stage). It also makes the writes (4) a tiny bit slower, since it's doing one to many writes now. But the benefit is that the binding (5) is straightforward: we now have just arrays of linear data that need to be mapped directly to Metal resource space, no extra work involved.

This trade-off makes sense because we expect the binding to occur more often than the earlier stages. And it does show the benefit on our benchmarks. It also opens the way for us to bind multiple descriptors at once (resource bundling), if we find it useful. There is a bit of an annoyance about handling the dynamic buffer offsets (yes, those again...) but this looks nicer with resource bundles.

File: src/backend/metal/src/command.rs [hunk 1]
```diff
@@ -2944,100 +2951,65 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
+                            let out = &mut cache.textures[index as usize];
+                            if *out != texture {
+                                *out = texture;
+                                pre.issue(soft::RenderCommand::BindTexture { stage, index, texture });
                             }
                         }
+                        for (index, &buffer) in (res_offset.buffers .. )
+                            .zip(&data.buffers[pool_offset.buffers.start as usize .. pool_offset.buffers.end as usize])
+                        {
+                            cache.buffers[index as usize] = buffer;
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: Is there any risk of `cache.buffers` becoming stale when `pre.is_void()` occurs (i.e. we don't submit `BindBuffer`)? What's the reason only `BindBuffer` is optional in that case?

Comment (grovesNL) [reply src/backend/metal/src/command.rs]: So wouldn't `if !pre.is_void() { pre.issue(soft::RenderCommand::BindSampler { stage, index, sampler }); }` be valid, for example?

**Comment by kvark** [comment]
It would be valid but useless, since issue() checks for void anyway

File: src/backend/metal/src/command.rs (hunk 1 above)
Comment (kvark) [reply src/backend/metal/src/command.rs]: all of the `is_void` checks we have are solely for optimization, they aren't functional. It's just to spare some work done for producing the commands that are passed to `pre.issue`

**Pull request merged**